### PR TITLE
peltool: Add --list option for PEL summary

### PR DESCRIPTION
**Pull request by harsh-agarwal1**
This commit introduces the --list option.
To retrieve a summary of PELs present in the specified input directory. The default directory is "/var/lib/phosphor-logging/extensions/pels/logs/".

Signed-off-by: harsh-agarwal1 Harsh.Agarwal@ibm.com

### CODE REVIEW

**Comment by spinler** [comment]
Can you show some example output?  In an issue comment is fine.

File: modules/pel/peltool/peltool.py [hunk 1]
```diff
@@ -267,6 +319,8 @@ def main():
                         help='Used with -d, only look for files with this extension (e.g. ".pel")')
     parser.add_argument('-x', '--delete', dest='delete', action='store_true',
                         help='Delete original file after parsing')
+    parser.add_argument('-l', '--list', nargs='?', const="/var/lib/phosphor-logging/extensions/pels/logs/", 
+                        default=None, help='Get summary of pels present in the input directory')
```
Comment (spinler) [inline modules/pel/peltool/peltool.py:323]: Let's keep the help the same as with the other peltool:  List PELs

File: modules/pel/peltool/peltool.py [hunk 2]
```diff
@@ -305,6 +359,23 @@ def main():
...
+        for root, _, files in os.walk(args.list):
+            for file in files:
+                if args.extension and args.extension != os.path.splitext(file)[1]:
+                    continue
+                eid, summary = parseAndSummarise(os.path.join(root, file), config, args.delete)
+                if(eid):
+                    final_summary[eid] = summary
+            # Only process top level directory
+            break
+        print(json.dumps(final_summary, indent=4))
```
Comment (spinler) [inline modules/pel/peltool/peltool.py:376]: These would need to be sorted based on the commit timestamp field since EIDs so that EIDs from different subsystems aren't grouped together.

PELs on the BMC have the commit timestamp as part of the filename, and it's also in the private header.

File: modules/pel/peltool/peltool.py [hunk 3]
```diff
@@ -267,6 +319,8 @@ def main():
                         help='Used with -d, only look for files with this extension (e.g. ".pel")')
     parser.add_argument('-x', '--delete', dest='delete', action='store_true',
                         help='Delete original file after parsing')
+    parser.add_argument('-l', '--list', nargs='?', const="/var/lib/phosphor-logging/extensions/pels/logs/",
```
Comment (spinler) [inline modules/pel/peltool/peltool.py:322]: For now, I don't want to expose these arguments to people running peltool when they aren't on the BMC.  I think we could get away which checking if we're on the BMC by looking to see if a certain directory is present, like either /usr/share/phosphor-logging or /var/lib/phosphor-logging.

After everything is in and working, and then we can reconsider if we'd want to show some or all of them.

File: modules/pel/peltool/peltool.py [hunk 4]
```diff
@@ -307,6 +307,23 @@ def main():
...
         sys.exit(0)
 
+    if args.list:
+        if not os.path.isdir(args.list):
+            sys.exit(f"{args.list} is not a valid directory")
+        print("This function is under development")
+        final_summary={}
+        for root, _, files in os.walk(args.list):
+            for file in files:
+                if args.extension and args.extension != os.path.splitext(file)[1]:
```
Comment (RameshIyyar) [inline modules/pel/peltool/peltool.py:369]: are we conveying this extension argument details as part of the list option usage? or could we use python argparse magic?

File: modules/pel/peltool/peltool.py [hunk 5]
```diff
@@ -307,6 +307,23 @@ def main():
...
+                    continue
+                eid, summary = parseAndSummarise(os.path.join(root, file), config, args.delete)
+                if(eid):
+                    final_summary[eid] = summary
```
Comment (RameshIyyar) [inline modules/pel/peltool/peltool.py:373]: do we have guarantee that PEL id won't duplicate? or do we need to indicate some error message?

File: modules/pel/peltool/peltool.py [hunk 6]
```diff
@@ -245,6 +245,22 @@ def parseAndWriteOutput(file: str, output_dir: str, config: Config,
             print(f"No PEL parsed for {file}: {e}")
 
 
+def parseAndSummarise(file: str, config: Config,
```
Comment (RameshIyyar) [inline modules/pel/peltool/peltool.py:284]: Please add function documentation.

I think you function name is confusing `parseAndSummarise` and `parsePelSummary` because you are just processing only one file, right?

I think you are reading first and parsing so please use name respectively.

File: modules/pel/peltool/peltool.py [hunk 7]
```diff
@@ -245,6 +245,22 @@ def parseAndWriteOutput(file: str, output_dir: str, config: Config,
...
+                        delete_after_parsing: bool):
+    with open(file, 'rb') as fd:
+        data = fd.read()
+        stream = DataStream(data, byte_order='big', is_signed=False)
```
Comment (RameshIyyar) [inline modules/pel/peltool/peltool.py:288]: I think endianness will be a issue if we are trying to use this tool in the BMC because BMC is little endian.

we need to check the system endianness and read it?

Actually I think it should based on the PEL file writer, not based on the system from where we are running the PEL tool. @spinler any thoughts?

File: modules/pel/peltool/peltool.py [hunk 8]
```diff
@@ -245,6 +245,22 @@ def parseAndWriteOutput(file: str, output_dir: str, config: Config,
...
+        try:
+            eid, summary = parsePelSummary(stream, config)
+            if(eid):
+                return eid, summary
+            else:
+                print(f"No PEL parsed for {file}")
```
Comment (RameshIyyar) [inline modules/pel/peltool/peltool.py:294]: How de we get error message and PEL summary both will be printed in the console, right?

Please add test result with negative case as well to check command output.

File: modules/pel/peltool/peltool.py [hunk 9]
```diff
@@ -245,6 +245,22 @@ def parseAndWriteOutput(file: str, output_dir: str, config: Config,
...
+        except Exception as e:
+            print(f"No PEL parsed for {file}: {e}")
```
Comment (RameshIyyar) [inline modules/pel/peltool/peltool.py:296]: Suggestion: Good to add "Exception" as prefix in the trace if "e" is just print exception message.

File: modules/pel/peltool/peltool.py [hunk 10]
```diff
@@ -245,6 +245,42 @@ def parseAndWriteOutput(file: str, output_dir: str, config: Config,
             print(f"No PEL parsed for {file}: {e}")
```
Comment (RameshIyyar) [inline modules/pel/peltool/peltool.py:245]: I feel, commit 1 and 2 could be squashed and then commit 3 and 4 could be squashed.

File: modules/pel/peltool/peltool.py [hunk 11]
```diff
@@ -245,6 +245,42 @@ def parseAndWriteOutput(file: str, output_dir: str, config: Config,
...
 
 
+def parsePelSummary(stream: DataStream, config: Config):
```
Comment (RameshIyyar) [inline modules/pel/peltool/peltool.py:248]: Please add function documentation.

File: modules/pel/peltool/peltool.py [hunk 12]
```diff
@@ -245,6 +245,42 @@ def parseAndWriteOutput(file: str, output_dir: str, config: Config,
...
+    out = OrderedDict()
+    ret, ph = generatePH(stream, out)
+    if ret is False:
+        return "", ""
+    eid = ph.pLID
+    ret, uh = generateUH(stream, ph.creatorID, out)
+    if ret is False:
+        return "", ""
+    
+    if config.serviceable_only and not uh.isServiceable():
```
Comment (RameshIyyar) [inline modules/pel/peltool/peltool.py:258]: These info could be added into function documentation.

File: modules/pel/peltool/peltool.py [hunk 13]
```diff
@@ -245,6 +245,42 @@ def parseAndWriteOutput(file: str, output_dir: str, config: Config,
...
+        return "", ""
+
+    if config.non_serviceable_only and uh.isServiceable():
```
Comment (RameshIyyar) [inline modules/pel/peltool/peltool.py:261]: Same as LN 258

**Review by RameshIyyar — CHANGES_REQUESTED** [review]

File: modules/pel/peltool/peltool.py (hunk 3 above)
Comment (harsh-agarwal1) [reply modules/pel/peltool/peltool.py:322]: Yeah, we can use this option even if we are not in a BMC. 

With the current code, if a user doesn't enter anything with -l, it will take default location as mentioned in const. 
However, if the user adds a directory after -l, then that directory will be stored in args.list.

For example:
```bash
python3 -m pel.peltool.peltool -l
```
Here args.list will store "/var/lib/phosphor-logging/extensions/pels/logs/"
```bash 
python3 -m pel.peltool.peltool -l testfolder/ 
```
Here args.list will store "testfolder/"

File: modules/pel/peltool/peltool.py [hunk 14]
```diff
@@ -307,6 +307,23 @@ def main():
 
         sys.exit(0)
 
+    if args.list:
+        if not os.path.isdir(args.list):
```
[the remark this answers is not in the document]
Comment (harsh-agarwal1) [reply modules/pel/peltool/peltool.py:363]: If the user does not enter any directory location, then with current code it will take the default location. Which is that of BMC's PELs location. 
If we are not in BMC, or the entered directory by user doesn't exist, that could be a problem. Hence put a check before moving ahead.

File: modules/pel/peltool/peltool.py [hunk 15]
```diff
@@ -267,6 +267,8 @@ def main():
                         help='Used with -d, only look for files with this extension (e.g. ".pel")')
```
[the remark this answers is not in the document]
Comment (spinler) [reply modules/pel/peltool/peltool.py:319]: i don't want to add a -list for the standalone pel parser yet

File: modules/pel/peltool/peltool.py (hunk 12 above)
Comment (spinler) [reply modules/pel/peltool/peltool.py:258]: we really only need the serviceable_only stuff for the standalone parser.  we already have the hidden and info flags for the BMC pel parser.

File: modules/pel/peltool/peltool.py (hunk 13 above)
Comment (RameshIyyar) [reply modules/pel/peltool/peltool.py:261]: Yeah I understood, requesting to add some details in the function documentation what it will return. Its your wish too if you think code explain that.

File: modules/pel/peltool/peltool.py (hunk 3 above)
Comment (harsh-agarwal1) [reply modules/pel/peltool/peltool.py:322]: Got it!

File: modules/pel/peltool/peltool.py (hunk 7 above)
Comment (spinler) [reply modules/pel/peltool/peltool.py:288]: PEL data is bit endian, and that is what the arg on DataStream is specifying.

File: modules/pel/peltool/peltool.py (hunk 3 above)
Comment (spinler) [reply modules/pel/peltool/peltool.py:322]: I should add, the same goes with all of these new options.

Comment (spinler) [reply modules/pel/peltool/peltool.py:322]: If we were going to show -l on non-BMC, it would need to detect that and require an argument.

File: modules/pel/peltool/peltool.py (hunk 1 above)
Comment (harsh-agarwal1) [reply modules/pel/peltool/peltool.py]: Resolved!

File: modules/pel/peltool/peltool.py (hunk 2 above)
Comment (harsh-agarwal1) [reply modules/pel/peltool/peltool.py]: Resolved!

File: modules/pel/peltool/peltool.py (hunk 3 above)
Comment (harsh-agarwal1) [reply modules/pel/peltool/peltool.py]: Resolved!
In a non-BMC it will simply ask the user for a folder path.

File: modules/pel/peltool/peltool.py [hunk 16]
```diff
@@ -307,6 +307,23 @@ def main():
 
         sys.exit(0)
 
+    if args.list:
```
[the remark this answers is not in the document]
Comment (harsh-agarwal1) [reply modules/pel/peltool/peltool.py:383]: Resolved!

**Review by harsh-agarwal1 — COMMENTED** [review]
Reviewed all the comments and addressed them all. Please have a look at the latest commits.

File: modules/pel/peltool/peltool.py (hunk 10 above)
Comment (spinler) [reply modules/pel/peltool/peltool.py:245]: yea, lets get the 2 commits down to 1.

File: modules/pel/peltool/peltool.py [hunk 17]
```diff
@@ -305,6 +380,15 @@ def main():
...
+        path = "/var/lib/phosphor-logging/extensions/pels/logs/"
+        if not os.path.isdir(path):
+            path = input("Enter folder path: ")
```
Comment (spinler) [inline modules/pel/peltool/peltool.py:386]: Oh, I meant check if on a BMC or not before doing the add_argument so you can handle it differently.  anything we do either way needs to be scriptable so can't ask extra questions.

**Comment by spinler** [comment]
Not that github allows me to add comments to commit messages :(, but there needs to be an empty line in between the message and description or else git mangles it:

```
Author: harsh-agarwal1 <Harsh.Agarwal@ibm.com>
Date:   Mon Apr 8 10:54:31 2024 +0530

    peltool: Implement function for PEL parsing and summarization
    Added function that would return EID and summary of a
    PEL file.
    This is building block for --list option
```
```
$ git log --oneline -1 d4325ba
d4325ba peltool: Implement function for PEL parsing and summarization Added function that would return EID and summary of a PEL file. This is building block for --list option
```

File: modules/pel/peltool/peltool.py [hunk 18]
```diff
@@ -245,6 +245,78 @@ def parseAndWriteOutput(file: str, output_dir: str, config: Config,
...
+    
+    summary = OrderedDict()
+    for _ in range(2, ph.sectionCount):
+        sectionID, sectionLen, versionID, subType, componentID = parseHeader(stream)
+        section_json = OrderedDict()
+        sectionFun(stream, section_json, sectionID, sectionLen,
+                   versionID, subType, componentID, ph.creatorID, config)
+        if sectionID == SectionID.primarySRC.value or sectionID == SectionID.secondarySRC.value:
+            summary["SRC"] = section_json["Primary SRC"]["Reference Code"]
+            if("Error Details" in section_json["Primary SRC"]):
```
Comment (spinler) [inline modules/pel/peltool/peltool.py:273]: please remove the ().  Really I only noticed because there was no space.

File: modules/pel/peltool/peltool.py [hunk 19]
```diff
@@ -245,6 +245,78 @@ def parseAndWriteOutput(file: str, output_dir: str, config: Config,
...
+    Extracts and summarizes data from a PEL file.
+    Returns: Event ID (eid) and summary extracted from the PEL.
+            If no PEL is parsed, empty strings are returned.
+    """
+    with open(file, 'rb') as fd:
+        data = fd.read()
+        stream = DataStream(data, byte_order='big', is_signed=False)
+        try:
+            eid, summary = parsePELSummary(stream, config)
+            if(eid):
```
Comment (spinler) [inline modules/pel/peltool/peltool.py:296]: remove the ()

File: modules/pel/peltool/peltool.py [hunk 20]
```diff
@@ -245,6 +245,78 @@ def parseAndWriteOutput(file: str, output_dir: str, config: Config,
...
+            if extension and extension != os.path.splitext(file)[1]:
+                continue
+            file_list.append(file)  ## create list of file names
+        # Only process top level directory
+        break
+    file_list.sort(reverse=False) 
+    final_summary = {}
+    for file in file_list:
+        eid, summary = extractAndSummarizePEL(os.path.join(root, file), config, delete)
+        if(eid):
```
Comment (spinler) [inline modules/pel/peltool/peltool.py:315]: remove the ()

**Comment by harsh-agarwal1** [comment]
Reviewed and addressed comments. Please have a look at the latest commits.

File: modules/pel/peltool/peltool.py [hunk 21]
```diff
@@ -245,6 +245,78 @@ def parseAndWriteOutput(file: str, output_dir: str, config: Config,
...
+    """
+    out = OrderedDict()
+    ret, ph = generatePH(stream, out)
+    if ret is False:
+        return "", ""
+    eid = ph.pLID
+    ret, uh = generateUH(stream, ph.creatorID, out)
+    if ret is False:
+        return "", ""
+    if not uh.isServiceable():
```
Comment (spinler) [inline modules/pel/peltool/peltool.py:262]: the C++ peltool also skips if the hidden action flag is set.  Usually hidden is always set on recovered or info PELs, so I think this is probably good enough, just something we need to keep an eye on.

File: modules/pel/peltool/peltool.py [hunk 22]
```diff
@@ -245,6 +245,78 @@ def parseAndWriteOutput(file: str, output_dir: str, config: Config,
...
+        return "", ""
+    
+    summary = OrderedDict()
+    for _ in range(2, ph.sectionCount):
+        sectionID, sectionLen, versionID, subType, componentID = parseHeader(stream)
+        section_json = OrderedDict()
+        sectionFun(stream, section_json, sectionID, sectionLen,
+                   versionID, subType, componentID, ph.creatorID, config)
+        if sectionID == SectionID.primarySRC.value or sectionID == SectionID.secondarySRC.value:
```
Comment (spinler) [inline modules/pel/peltool/peltool.py:271]: you can remove the SecondarySRC check here.  in this case, we'll just want to show the summary for the primary SRC section

File: modules/pel/peltool/peltool.py [hunk 23]
```diff
@@ -245,6 +245,78 @@ def parseAndWriteOutput(file: str, output_dir: str, config: Config,
...
+            file_list.append(file)  ## create list of file names
+        # Only process top level directory
+        break
+    file_list.sort(reverse=False) 
+    final_summary = {}
+    for file in file_list:
+        eid, summary = extractAndSummarizePEL(os.path.join(root, file), config, delete)
+        if eid :
+            final_summary[eid] = summary
+    print(json.dumps(final_summary, indent=4))
```
Comment (spinler) [inline modules/pel/peltool/peltool.py:317]: now I think you can rebase on master and call prettyPrint here?

**Comment by harsh-agarwal1** [comment]
Reviewed and addressed comments. Please have a look at the latest commit.

File: modules/pel/peltool/peltool.py [hunk 24]
```diff
@@ -327,6 +406,10 @@ def main():
 
         sys.exit(0)
 
+    if inBMC and args.list:
+        listOption("/var/lib/phosphor-logging/extensions/pels/logs/", config, args.extension, args.delete)
```
Comment (RameshIyyar) [inline modules/pel/peltool/peltool.py:410]: Please use some constant, the same value (path) we hardcoded more than one places so it will be difficult if something (path) is changed in future.

File: modules/pel/peltool/peltool.py [hunk 25]
```diff
@@ -267,7 +267,82 @@ def parseAndWriteOutput(file: str, output_dir: str, config: Config,
...
+    summary["PLID"] = ph.pLID
+    summary["CreatorID"] = out["Private Header"]["Creator Subsystem"]
+    summary["Subsystem"] = out["User Header"]["Subsystem"]
+    summary["Commit Time"] = ph.commitTime
+    summary["Sev"] = out["User Header"]["Event Severity"]
+    summary["CompID"] = out["Private Header"]["Created by"]
+    return eid, summary
+
+def extractAndSummarizePEL(file: str, config: Config,
+                        delete_after_parsing: bool):
```
Comment (RameshIyyar) [inline modules/pel/peltool/peltool.py:307]: `delete_after_parsing` is unused, are you defining unwanted arguments and passing through all functions from `main` function.

File: modules/pel/peltool/peltool.py [hunk 26]
```diff
@@ -267,7 +267,82 @@ def parseAndWriteOutput(file: str, output_dir: str, config: Config,
             print(f"No PEL parsed for {file}: {e}")
 
 
+def parsePELSummary(stream: DataStream, config: Config, critSysTerm:bool = False):
```
Comment (RameshIyyar) [inline modules/pel/peltool/peltool.py:270]: Same as mentioned in LN307, you are not using `critSysTerm`, right?

File: modules/pel/peltool/peltool.py [hunk 27]
```diff
@@ -267,7 +267,82 @@ def parseAndWriteOutput(file: str, output_dir: str, config: Config,
...
+    """
+    with open(file, 'rb') as fd:
+        data = fd.read()
+        stream = DataStream(data, byte_order='big', is_signed=False)
+        try:
+            eid, summary = parsePELSummary(stream, config)
+            if eid :
+                return eid, summary
+        except Exception as e:
+            print(f"Exception: No PEL parsed for {file}: {e}")
```
Comment (RameshIyyar) [inline modules/pel/peltool/peltool.py:321]: Do you have any test result for the negative case, I would like to make sure how are we printing error details if some PEL is not able to parse and PEL summary of few PELs if able to parse.

File: modules/pel/peltool/peltool.py [hunk 28]
```diff
@@ -212,7 +212,7 @@ def parsePEL(stream: DataStream, config: Config, exit_on_error: bool):
             return "", ""
```
Comment (RameshIyyar) [inline modules/pel/peltool/peltool.py:212]: Commit Message:

Please add new line after commit message header.

File: modules/pel/peltool/peltool.py (hunk 25 above)
Comment (harsh-agarwal1) [reply modules/pel/peltool/peltool.py:307]: Currently not using these extra arguments, will be useful for future feature implementation.

File: modules/pel/peltool/peltool.py (hunk 27 above)
Comment (harsh-agarwal1) [reply modules/pel/peltool/peltool.py:321]: Yeah, For a file with text content:  `PH test test` , 
Got this before full output:

 `Exception: No PEL parsed for testfolder/errp: range check failure`

File: modules/pel/peltool/peltool.py (hunk 25 above)
Comment (RameshIyyar) [reply modules/pel/peltool/peltool.py:307]: Its ok for now but, we should push into relevant patch..Consider this practise in future.

File: modules/pel/peltool/peltool.py (hunk 27 above)
Comment (RameshIyyar) [reply modules/pel/peltool/peltool.py:321]: @spinler are you ok if we printed errors PELs?

File: modules/pel/peltool/peltool.py [hunk 29]
```diff
@@ -267,7 +267,83 @@ def parseAndWriteOutput(file: str, output_dir: str, config: Config,
...
+        for file in files:
+            if extension and extension != os.path.splitext(file)[1]:
+                continue
+            file_list.append(file)  ## create list of file names
+        # Only process top level directory
+        break
+    file_list.sort(reverse=rev) 
+    final_summary = {}
+    for file in file_list:
+        eid, summary = extractAndSummarizePEL(os.path.join(root, file), config, delete)
```
Comment (spinler) [inline modules/pel/peltool/peltool.py:336]: pyright gives a warning here: "root" is possibly unbound

File: modules/pel/peltool/peltool.py (hunk 27 above)
Comment (spinler) [reply modules/pel/peltool/peltool.py:321]: Ideally no, though for example if that PH wasn't there some other code would print an error already.
Theoretically this shouldn't happen because the PEL daemon will already delete bad PELs, but who knows.

What if we added a --verbose option and then print the error here if set?  I think that wouldn't be too bad to add?  just set a new flag on Config?

File: modules/pel/peltool/peltool.py [hunk 30]
```diff
@@ -289,6 +365,10 @@ def main():
                         help='Used with -d, only look for files with this extension (e.g. ".pel")')
     parser.add_argument('-x', '--delete', dest='delete', action='store_true',
                         help='Delete original file after parsing')
+    if inBMC:
+        parser.add_argument('-l', '--list',
+                            help='List PELs',
+                            action='store_true')
```
Comment (spinler) [inline modules/pel/peltool/peltool.py:371]: I think I've come around to having a -l in the nonBMC case, since now it's as easy as:

```
else:
           parser.add_argument('-l', '--list',
                            help='List PELs',
                            dest=list_directory')
  
          # if not isdir(list_dir) fail ...
```
does that seem like it would work?

File: modules/pel/peltool/peltool.py (hunk 29 above)
Comment (harsh-agarwal1) [reply modules/pel/peltool/peltool.py:337]: The root represents the current directory being traversed. However, in this case, pyright must be warning that root might not assigned a value.
Added default value to root, this should resolve the warning.

File: modules/pel/peltool/peltool.py (hunk 30 above)
Comment (harsh-agarwal1) [reply modules/pel/peltool/peltool.py:372]: Added the functionality.

File: modules/pel/peltool/peltool.py (hunk 27 above)
Comment (harsh-agarwal1) [reply modules/pel/peltool/peltool.py:321]: If we add --verbose, we will also want to put all other error message under this. 
Like [LN42](https://github.com/ibm-openbmc/openpower-pel-parsers/blob/master/modules/pel/peltool/peltool.py#L42), LN55, etc.
This could be a functionality in itself.

Comment (harsh-agarwal1) [reply modules/pel/peltool/peltool.py:321]: We could handle it separately for the whole tool by adding verbose option.

File: modules/pel/peltool/peltool.py [hunk 31]
```diff
@@ -327,6 +412,17 @@ def main():
 
         sys.exit(0)
 
+    if args.list:
+        if inBMC:
+            listOption(BMCpath, config, args.extension, args.delete)
+            sys.exit(0)
+        else:
+            if not os.path.isdir(args.list):
```
Comment (RameshIyyar) [inline modules/pel/peltool/peltool.py:420]: Suggestion/FYI: we could use https://docs.python.org/3/library/argparse.html#action to validate the user inputs.

Ignore for now.

File: modules/pel/peltool/peltool.py [hunk 32]
```diff
@@ -267,7 +267,84 @@ def parseAndWriteOutput(file: str, output_dir: str, config: Config,
...
+    for file in file_list:
+        eid, summary = extractAndSummarizePEL(os.path.join(root, file), config, delete)
+        if eid :
+            final_summary[eid] = summary
+    print(prettyPrint(json.dumps(final_summary, indent=4)))
+
+
 def main():
+    inBMC=False
+    BMCpath="/var/lib/phosphor-logging/extensions/pels/logs/"
```
Comment (RameshIyyar) [inline modules/pel/peltool/peltool.py:345]: nit: BMCPath will confuse, BMCPELsPath.

File: modules/pel/peltool/peltool.py [hunk 33]
```diff
@@ -289,6 +366,14 @@ def main():
...
     parser.add_argument('-x', '--delete', dest='delete', action='store_true',
                         help='Delete original file after parsing')
+    if inBMC:
+        parser.add_argument('-l', '--list',
+                            help='List PELs',
+                            action='store_true')
+    else:
+        parser.add_argument('-l', '--list',
+                            help='List PELs',
+                            dest='list')
```
Comment (RameshIyyar) [inline modules/pel/peltool/peltool.py:376]: Suggestion/FYI: I would use https://docs.python.org/3/library/argparse.html#metavar to inform about the directory arguments.

**Review by RameshIyyar — APPROVED** [review]

File: modules/pel/peltool/peltool.py (hunk 31 above)
Comment (harsh-agarwal1) [reply modules/pel/peltool/peltool.py:420]: Currently using action as required, for example in [LN372](https://github.com/ibm-openbmc/openpower-pel-parsers/commit/1fec449f71217319e6f78fa73ab5657b6f0273a2#diff-f91c747e8d1a325f39181d35322d5c5d6f146b18af0b10b74dd0657e1da155cbR372). Will improve the use in future.

File: modules/pel/peltool/peltool.py (hunk 33 above)
Comment (harsh-agarwal1) [reply modules/pel/peltool/peltool.py:376]: Noted!

File: modules/pel/peltool/peltool.py [hunk 34]
```diff
@@ -267,7 +267,84 @@ def parseAndWriteOutput(file: str, output_dir: str, config: Config,
...
+            file_list.append(file)  ## create list of file names
+        # Only process top level directory
+        break
+    file_list.sort(reverse=rev) 
+    final_summary = {}
+    for file in file_list:
+        eid, summary = extractAndSummarizePEL(os.path.join(root, file), config, delete)
+        if eid :
+            final_summary[eid] = summary
+    print(prettyPrint(json.dumps(final_summary, indent=4)))
```
Comment (spinler) [inline modules/pel/peltool/peltool.py:340]: there's a bit more of an indent with this than with the original peltool:

original:
```
"SRC":                  "110015F0",
"Message":              "A power suppl
```
this one:
```
"SRC":                       "BD503001",
"Message":                   "Failure occurred during boot process",
```

not a huge deal but is there a way to remove some spaces (looks like its 5) from this output?

File: modules/pel/peltool/peltool.py (hunk 29 above)
Comment (spinler) [reply modules/pel/peltool/peltool.py:337]: yep, that fixed it.  thanks.

File: modules/pel/peltool/peltool.py (hunk 34 above)
Comment (spinler) [reply modules/pel/peltool/peltool.py]: looks good!

File: modules/pel/peltool/peltool.py [hunk 35]
```diff
@@ -327,6 +412,17 @@ def main():
 
         sys.exit(0)
 
+    if args.list:
+        if inBMC:
+            listOption(BMCPELsPath, config, args.extension, args.delete)
+            sys.exit(0)
+        else:
+            if not os.path.isdir(args.list):
+                print("Error: Given path is not a directory")
```
Comment (spinler) [inline modules/pel/peltool/peltool.py:421]: sorry, one minor thing:  to be consistent, lets exit how it was done on line 307 and use the sys.exit() option that takes a string.

also, you can move the sys.exit(0) to the bottom of the if args.list block and then just have it once
